test/functional/main_test.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"context"
+	"flag"
+	"log"
+	"os"
+	"os/exec"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/hcsoci"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+
+	testctrd "github.com/Microsoft/hcsshim/test/internal/containerd"
+	testflag "github.com/Microsoft/hcsshim/test/internal/flag"
+	"github.com/Microsoft/hcsshim/test/internal/require"
+)
+
+// owner field for uVMs
+const hcsOwner = "hcsshim-functional-tests"
+
+const (
+	featureLCOW        = "LCOW"
+	featureWCOW        = "WCOW"
+	featureContainer   = "container"
+	featureHostProcess = "HostProcess"
+	featureUVMMem      = "UVMMem"
+	featurePlan9       = "Plan9"
+	featureSCSI        = "SCSI"
+	featureScratch     = "Scratch"
+	featureVSMB        = "vSMB"
+	featureVPMEM       = "vPMEM"
+)
+
+var allFeatures = []string{
+	featureLCOW,
+	featureWCOW,
+	featureHostProcess,
+	featureContainer,
+	featureUVMMem,
+	featurePlan9,
+	featureSCSI,
+	featureScratch,
+	featureVSMB,
+	featureVPMEM,
+}
+
+// todo: use a new containerd namespace and then nuke everything in it
+
+var (
+	debug                                 bool
+	pauseDurationOnCreateContainerFailure time.Duration
+
+	// flags
+	flagFeatures            = testflag.NewFeatureFlag(allFeatures)
+	flagContainerdAddress   = flag.String("ctr-address", "tcp://127.0.0.1:2376", "Address for containerd's GRPC server")
+	flagContainerdNamespace = flag.String("ctr-namespace", "k8s.io", "Containerd namespace")
+	flagCtrExePath          = flag.String("ctr-path", `C:\ContainerPlat\ctr.exe`, "Path to ctr.exe")
+	flagLinuxBootFilesPath  = flag.String("linux-bootfiles",
+		`C:\\ContainerPlat\\LinuxBootFiles`,
+		"Path to LCOW UVM boot files (rootfs.vhd, initrd.img, kernel, vmlinux)")
+)
+
+func init() {
+	if !winapi.IsElevated() {
+		log.Fatal("tests must be run in an elevated context")
+	}
+
+	if _, ok := os.LookupEnv("HCSSHIM_FUNCTIONAL_TESTS_DEBUG"); ok {
+		debug = true
+	}
+	flag.BoolVar(&debug, "debug", debug, "Set logging level to debug [%HCSSHIM_FUNCTIONAL_TESTS_DEBUG%]")
+
+	// This allows for debugging a utility VM.
+	if s := os.Getenv("HCSSHIM_FUNCTIONAL_TESTS_PAUSE_ON_CREATECONTAINER_FAIL_IN_MINUTES"); s != "" {
+		if t, err := strconv.Atoi(s); err == nil {
+			pauseDurationOnCreateContainerFailure = time.Duration(t) * time.Minute
+		}
+	}
+	flag.DurationVar(&pauseDurationOnCreateContainerFailure,
+		"container-creation-failure-pause",
+		pauseDurationOnCreateContainerFailure,
+		"The number of minutes to wait after a container creation failure to try again "+
+			"[%HCSSHIM_FUNCTIONAL_TESTS_PAUSE_ON_CREATECONTAINER_FAIL_IN_MINUTES%]")
+
+	// Try to stop any pre-existing compute processes
+	// cmd := exec.Command("powershell", `get-computeprocess | stop-computeprocess -force`)
+	// _ = cmd.Run()
```

Comment:
What was this for?

Reply:
the old tests would kill any running hyperv containers/vms
which kills wsl and other things
I added a line to clean up any lingering hyperv stuff started by this test, so I can delete that